firmware/user/modes/mode_mtype.c

```diff
@@ -0,0 +1,1053 @@
+
+    int reflectBarFillX1 = reflectBarX0 + (charge * ((reflectBarX1 - 1) - reflectBarX0));
+    plotLine(reflectBarX0, reflectBarY0 + 1, reflectBarFillX1, reflectBarY0 + 1, WHITE);
+
+    // For each chunk coordinate
+    for(uint8_t w = 0; w < NUM_CHUNKS + 1; w++)
+    {
+        // Plot a floor segment line between chunk coordinates
+        plotLine(
+            (w * CHUNK_WIDTH) - mType->xOffset,
+            mType->floors[w],
+            ((w + 1) * CHUNK_WIDTH) - mType->xOffset,
+            mType->floors[w + 1],
+            WHITE);
+    }
+
+    /*char uiStr[32] = {0};
+    ets_snprintf(uiStr, sizeof(uiStr), "BTN: %d u:%d, d:%d, l:%d, r:%d, a:%d", button, upDown, downDown, leftDown, rightDown, actionDown);
+    fillDisplayArea(OLED_WIDTH - getTextWidth(uiStr, TOM_THUMB) - 1, OLED_HEIGHT - (1 * (FONT_HEIGHT_TOMTHUMB + 1)), OLED_WIDTH, OLED_HEIGHT, BLACK);
+    plotText(OLED_WIDTH - getTextWidth(uiStr, TOM_THUMB) - 1, OLED_HEIGHT - (1 * (FONT_HEIGHT_TOMTHUMB + 1)), uiStr, TOM_THUMB, WHITE);*/
+    
+}
+
+void ICACHE_FLASH_ATTR mtScoresInput(void)
+{
+
+}
+
+void ICACHE_FLASH_ATTR mtScoresLogic(void)
+{
+
+}
+
+void ICACHE_FLASH_ATTR mtScoresDisplay(void)
+{
+
+}
+
+void ICACHE_FLASH_ATTR mtGameoverInput(void)
+{
+
+}
+
+void ICACHE_FLASH_ATTR mtGameoverLogic(void)
+{
+
+}
+
+void ICACHE_FLASH_ATTR mtGameoverDisplay(void)
+{
+
+}
+
+/**
+ * TODO
+ *
+ * @param menuItem
+ */
+static void ICACHE_FLASH_ATTR mtMenuCallback(const char* menuItem)
+{
+    if (mt_start == menuItem)
+    {
+        // Change state to start game.
+        mtSetState(MT_GAME);
+    }
+    else if (mt_scores == menuItem)
+    {
+        // Change state to score screen.
+        mtSetState(MT_SCORES);
+    }
+    else if (mt_quit == menuItem)
+    {
+        // Exit this swadge mode.
+        switchToSwadgeMode(0);
+    }
+}
+
+uint8_t ICACHE_FLASH_ATTR getTextWidth(char* text, fonts font)
+{
+    // NOTE: The inverse, inverse is cute, but 2 draw calls, could we draw it outside of the display area but still in bounds of a uint8_t?
+
+    // We only get width info once we've drawn.
+    // So we draw the text as inverse to get the width.
+    uint8_t textWidth = plotText(0, 0, text, font,
+                                 INVERSE) - 1; // minus one accounts for the return being where the cursor is.
+
+    // Then we draw the inverse back over it to restore it.
+    plotText(0, 0, text, font, INVERSE);
+
+    return textWidth;
+}
+
+bool AABBCollision (int ax0, int ay0, int ax1, int ay1, int bx0, int by0, int bx1, int by1) {
+    int awidth = ax1 - ax0;
+    int aheight = ay1 - ay0;
+
+    int bwidth = bx1 - bx0;
+    int bheight = by1 - by0;
+
+    return (ax0 < bx1 &&
+            ax1 > bx0 &&
+            ay0 < by1 &&
+            ay1 > by0);
+}
+
+void normalize (vecdouble_t * vec)
```

Comment:
The raycaster required vector math too, and I ended up doing it without `pow()` or `sqrt()`. It's too late now, but a vector math util file would be nice for next year...
* https://github.com/AEFeinstein/Super-2021-Swadge-FW-Sandbox/blob/a3b5143dc4ef2431dbaf6b166900414901356c64/firmware/user/modes/mode_raycaster.c#L1407
* https://en.wikipedia.org/wiki/Fast_inverse_square_root